llama.cpp

```diff
@@ -5024,56 +5127,80 @@ static bool llm_load_tensors(
 
     ml.done_getting_tensors();
 
-    ml.init_mapping(true, use_mlock ? &model.mlock_mmap : nullptr);
+    ml.init_mappings(true, &model.mlock_mmaps);
 
     // create the backend buffers
-    std::vector<std::pair<ggml_context *, ggml_backend_buffer_t>> ctx_bufs;
+    std::vector<std::pair<ggml_context *, std::vector<ggml_backend_buffer_t>>> ctx_bufs;
 
     for (auto & it : ctx_map) {
         ggml_backend_buffer_type_t buft = it.first;
         ggml_context * ctx = it.second;
-        ggml_backend_buffer_t buf = nullptr;
+        std::vector<ggml_backend_buffer_t> bufs;
 
         // only the mmap region containing the tensors in the model is mapped to the backend buffer
         // this is important for metal with apple silicon: if the entire model could be mapped to a metal buffer, then we could just use metal for all layers
         // this allows using partial offloading when the model size exceeds the metal buffer size, but not the RAM size
         if (ml.use_mmap && buft == llama_default_buffer_type_cpu(true)) {
-            size_t first, last;
-            ml.get_mapping_range(&first, &last, ctx);
-            buf = ggml_backend_cpu_buffer_from_ptr((char *) ml.mapping->addr + first, last - first);
+            for (uint32_t file_no = 0; file_no < ml.files.size(); file_no++) {
```

Comment:
I suspect that here we can do a bit more simple: firstly, we calculate the total size of device buffer needed (which is sum of buffers required by each file), then allocate only one big device buffer. That should work on CUDA, but I'm not sure if it's the same on Metal? (since I'm not sure if we use mmap to take advantage of unified memory - I don't use apple silicon so maybe I'm totally wrong)

Reply:
I think it already works this way for CUDA, Metal needs a different buffer for each file to support mmap with unified memory as you say, same as the CPU.